### ISSUE TITLE
Can't train on YCB-V dataset

### ISSUE THREAD

**Issue by nikonnext** [ybkscht/EfficientPose#23]
Thanks for your work, @ybkscht

I've spent preprocessed the YCB-V (Original synt. Training images) dataset, which I took from the link https://bop.felk.cvut.cz/datasets/
From 10 folders of 1k images, I got a dataset of 236 high-quality images, on which the target object overlap is minimal. My preprocessed YCB-V dataset turned out to be identical to Linemod-preprocessed. The debug.py script works fine on my YCBV-preprocessed dataset. However, during training, I do not receive any metrics, respectively, mAP, ADD, ADD-S are equal to zero.

```
mAP: 0.0000
ADD: 0.0000
ADD-S: 0.0000
5cm_5degree: 0.0000
TranslationErrorMean_in_mm: nan
TranslationErrorStd_in_mm: nan
RotationErrorMean_in_degree: nan
RotationErrorStd_in_degree: nan
2D-Projection: 0.0000
Summed_Translation_Rotation_Error: nan
ADD(-S): 0.0000
AveragePointDistanceMean_in_mm: nan
AveragePointDistanceStd_in_mm: nan
AverageSymmetricPointDistanceMean_in_mm: nan
AverageSymmetricPointDistanceStd_in_mm: nan
MixedAveragePointDistanceMean_in_mm: nan
MixedAveragePointDistanceStd_in_mm: nan

Epoch 00005: ADD did not improve from 0.00000
1790/1790 [==============================] - 564s 315ms/step - loss: 3.5037 - classification_loss: 0.8148 - regression_loss: 0.3623 - transformation_loss: 116.3291

```
Please tell me what the lack of metrics can be connected with in my case.

**Comment by ybkscht** [comment]
Hi @nikonnext ,

looking at the logs there seems to be a problem in the 2D detection part in the training process (mAP 0.0).
A possible reason for this is that the generator is not loading the object's 2D bounding boxes correctly.
You can check this with `debug.py` when using the additional arguments `--draw_2d-bboxes` and `--draw-class-names`.
If so, maybe there went something wrong loading the segmentation masks.
Please note that the generator calculates the 2D bounding boxes from the segmentation masks instead of using the provided values directly.
This is done in `generators/linemod.py` at line 430
```
mask = cv2.imread(mask_path)
annotations["bboxes"][0, :], _ = self.get_bbox_from_mask(mask)
```

I hope this helps.

Sincerely,
Yannick

**Comment by nikonnext** [comment]
Thank you, @ybkscht, 

i don't quite understand what i should see when running debug.py with parameters `--draw_2d-bboxes` and `--draw-class-names`. However, the result is practically the same if I run debug with linemod-preprocessed. I think the masks are loading correctly because if I run debug.py with `--annotations` the rendering of the cuboids is correct.

**Comment by ybkscht** [comment]
With `--draw_2d-bboxes` you should see an additional 2D bounding box in dark green.
The following images show an example image with and without the `--draw_2d-bboxes` argument.

![no_bbox_2d](https://user-images.githubusercontent.com/74176077/118981142-53eb2480-b97a-11eb-941b-7c94c636c8e2.jpg)
![with_bbox_2d](https://user-images.githubusercontent.com/74176077/118981147-5483bb00-b97a-11eb-9b43-ddb82b370d18.jpg)

If you don't see such a bbox then there is probably an issue at this part.
EfficientPose basically extends a 2D object detector to provide an additional 6D pose for each object and this 2D bounding box is the provided ground truth for the 2D detection part of EfficientPose (the cuboid is the GT of the 6D pose).
So if you have not GT here then the object detection part can't learn to detect your object in the first place.

**Comment by nikonnext** [comment]
@ybkscht , if i use `python debug.py --annotations --draw_2d-bboxes --draw-class-names linemod path-dataset/ --object-id 4` the result is as follows:
![image](https://user-images.githubusercontent.com/70474232/119009291-a7219f00-b99b-11eb-9c02-581ee925f3e4.png)

**Comment by nikonnext** [comment]
@ybkscht , for training i used `python train.py --phi 0 linemod /path_to_dataset/ --object-id 4`
Maybe I should have changed some parameters in the train.pu script?

**Comment by ybkscht** [comment]
Ok that seems to be fine.
I strongly recommend using the provided COCO pretrained weights for training (`--weights` argument in train.py).
Maybe this is the problem here because it can take quite a bit when training from scratch to see any results and the batch norm layer are frozen during training with default settings which can possibly prevent the network from learning anything when randomly initialized.

**Comment by nikonnext** [comment]
thank you, @ybkscht
Can you tell me where I can find pre-trained COCO weights to fit me?

**Comment by ybkscht** [comment]
You can download them from [here](https://drive.google.com/drive/folders/1n3opeOo2ko9GwC9G5_OkVF1sk0NVd3YY). The link is also in the README

**Comment by nikonnext** [comment]
Thank you so much, @ybkscht 
I trained on the YCB-V dataset, here are the results for the fifth epoch:
```
201 instances of class object with average precision: 0.9999
201 instances of class object with ADD accuracy: 0.0348
201 instances of class object with ADD-S-Accuracy: 0.3582
201 instances of class object with 5cm-5degree-Accuracy: 0.0199
class object with Translation Differences in mm: Mean: 34.5008 and Std: 26.6614
class object with Rotation Differences in degree: Mean: 58.4149 and Std: 43.1900
201 instances of class object with 2d-projection-Accuracy: 0.0050
201 instances of class object with ADD(-S)-Accuracy: 0.0348
class object with Transformed Point Distances in mm: Mean: 48.2418 and Std: 26.5071
class object with Transformed Symmetric Point Distances in mm: Mean: 19.8826 and Std: 16.9235
class object with Mixed Transformed Point Distances in mm: Mean: 48.2418 and Std: 26.5071
mAP: 0.9999
ADD: 0.0348
ADD-S: 0.3582
5cm_5degree: 0.0199
TranslationErrorMean_in_mm: 34.5008
TranslationErrorStd_in_mm: 26.6614
RotationErrorMean_in_degree: 58.4149
RotationErrorStd_in_degree: 43.1900
2D-Projection: 0.0050
Summed_Translation_Rotation_Error: 162.7671
ADD(-S): 0.0348
AveragePointDistanceMean_in_mm: 48.2418
AveragePointDistanceStd_in_mm: 26.5071
AverageSymmetricPointDistanceMean_in_mm: 19.8826
AverageSymmetricPointDistanceStd_in_mm: 16.9235
MixedAveragePointDistanceMean_in_mm: 48.2418
MixedAveragePointDistanceStd_in_mm: 26.5071

Epoch 00005: ADD improved from 0.02488 to 0.03483, saving model to checkpoints/20_05_2021_20_14_59/object_4/phi_0_linemod_best_ADD.h5
1790/1790 [==============================] - 552s 309ms/step - loss: 1.6186 - classification_loss: 0.2770 - regression_loss: 0.1118 - transformation_loss: 61.4888
```
inference_webcam.py script works fine.

**Comment by shanice-l** [comment]
> Thank you so much, @ybkscht
> I trained on the YCB-V dataset, here are the results for the fifth epoch:
> 
> ```
> 201 instances of class object with average precision: 0.9999
> 201 instances of class object with ADD accuracy: 0.0348
> 201 instances of class object with ADD-S-Accuracy: 0.3582
> 201 instances of class object with 5cm-5degree-Accuracy: 0.0199
> class object with Translation Differences in mm: Mean: 34.5008 and Std: 26.6614
> class object with Rotation Differences in degree: Mean: 58.4149 and Std: 43.1900
> 201 instances of class object with 2d-projection-Accuracy: 0.0050
> 201 instances of class object with ADD(-S)-Accuracy: 0.0348
> class object with Transformed Point Distances in mm: Mean: 48.2418 and Std: 26.5071
> class object with Transformed Symmetric Point Distances in mm: Mean: 19.8826 and Std: 16.9235
> class object with Mixed Transformed Point Distances in mm: Mean: 48.2418 and Std: 26.5071
> mAP: 0.9999
> ADD: 0.0348
> ADD-S: 0.3582
> 5cm_5degree: 0.0199
> TranslationErrorMean_in_mm: 34.5008
> TranslationErrorStd_in_mm: 26.6614
> RotationErrorMean_in_degree: 58.4149
> RotationErrorStd_in_degree: 43.1900
> 2D-Projection: 0.0050
> Summed_Translation_Rotation_Error: 162.7671
> ADD(-S): 0.0348
> AveragePointDistanceMean_in_mm: 48.2418
> AveragePointDistanceStd_in_mm: 26.5071
> AverageSymmetricPointDistanceMean_in_mm: 19.8826
> AverageSymmetricPointDistanceStd_in_mm: 16.9235
> MixedAveragePointDistanceMean_in_mm: 48.2418
> MixedAveragePointDistanceStd_in_mm: 26.5071
> 
> Epoch 00005: ADD improved from 0.02488 to 0.03483, saving model to checkpoints/20_05_2021_20_14_59/object_4/phi_0_linemod_best_ADD.h5
> 1790/1790 [==============================] - 552s 309ms/step - loss: 1.6186 - classification_loss: 0.2770 - regression_loss: 0.1118 - transformation_loss: 61.4888
> ```
> 
> inference_webcam.py script works fine.

Could you please share your final results on YCBV datasets? The results for the fifth epoch seem not very satisfying.

**Comment by KulikovNA** [comment]
@arezu94 , I'm sorry, I haven't been on github for a long time. If this is still relevant, then I achieve 60% ADD-S on a custom dataset for 26 epochs. Conditions under which I get these metrics: --phi 0, --batch_size 8, + use pretrained EfficientDet weights. I have 12k images in my dataset, 10k of which I use for training. All images are synthetic, image size is 1280x960. In addition, I conducted a small experiment and it turned out that for me the optimal scale is phi 2, where my results are better. I tried all scales from 0 to 5, with different batch_size loading, phi 2 is optimal for me. However, with pre-trained EfficientDet weights, I get metrics already at the first epoch. Perhaps you are making a mistake somewhere.

**Comment by satpalsr** [comment]
Hey @nikonnext & @ybkscht. I have a custom dataset with RGB images, masked images & 3d model. But how should I obtain the values for cam_R_m2c & cam_T_m2c?

**Comment by SilvesterYu** [comment]
> inference_webcam.py

Hi, may I know the command you used to achieve this result? I am also having the nan problem.

**Comment by SilvesterYu** [comment]
> Thank you so much, @ybkscht I trained on the YCB-V dataset, here are the results for the fifth epoch:
> 
> ```
> 201 instances of class object with average precision: 0.9999
> 201 instances of class object with ADD accuracy: 0.0348
> 201 instances of class object with ADD-S-Accuracy: 0.3582
> 201 instances of class object with 5cm-5degree-Accuracy: 0.0199
> class object with Translation Differences in mm: Mean: 34.5008 and Std: 26.6614
> class object with Rotation Differences in degree: Mean: 58.4149 and Std: 43.1900
> 201 instances of class object with 2d-projection-Accuracy: 0.0050
> 201 instances of class object with ADD(-S)-Accuracy: 0.0348
> class object with Transformed Point Distances in mm: Mean: 48.2418 and Std: 26.5071
> class object with Transformed Symmetric Point Distances in mm: Mean: 19.8826 and Std: 16.9235
> class object with Mixed Transformed Point Distances in mm: Mean: 48.2418 and Std: 26.5071
> mAP: 0.9999
> ADD: 0.0348
> ADD-S: 0.3582
> 5cm_5degree: 0.0199
> TranslationErrorMean_in_mm: 34.5008
> TranslationErrorStd_in_mm: 26.6614
> RotationErrorMean_in_degree: 58.4149
> RotationErrorStd_in_degree: 43.1900
> 2D-Projection: 0.0050
> Summed_Translation_Rotation_Error: 162.7671
> ADD(-S): 0.0348
> AveragePointDistanceMean_in_mm: 48.2418
> AveragePointDistanceStd_in_mm: 26.5071
> AverageSymmetricPointDistanceMean_in_mm: 19.8826
> AverageSymmetricPointDistanceStd_in_mm: 16.9235
> MixedAveragePointDistanceMean_in_mm: 48.2418
> MixedAveragePointDistanceStd_in_mm: 26.5071
> 
> Epoch 00005: ADD improved from 0.02488 to 0.03483, saving model to checkpoints/20_05_2021_20_14_59/object_4/phi_0_linemod_best_ADD.h5
> 1790/1790 [==============================] - 552s 309ms/step - loss: 1.6186 - classification_loss: 0.2770 - regression_loss: 0.1118 - transformation_loss: 61.4888
> ```
> 
> inference_webcam.py script works fine.

Hi, may I know the command you used to achieve this result? I am also having the nan problem.
